File: Difficulty/Hard/[145] Binary Tree Postorder Traversal/solution.cpp
/*
 * @lc app=leetcode id=145 lang=cpp
 *
 * [145] Binary Tree Postorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        if (!root) return vector<int>();
        vector<int> nodes;
        stack<TreeNode*> todo;
        TreeNode* last = nullptr;
        while (root || !todo.empty()) {
            if (root) {
                todo.push(root);
                root = root -> left;
            } else {
                TreeNode* node = todo.top();
                if (node -> right && node -> right != last) {
                    root = node -> right;
                } else {
                    nodes.push_back(node->val);
                    last = node;
                    todo.pop();
                }
            }
        }
        return nodes;
    }
};
// @lc code=end

File: Difficulty/Hard/[145] Binary Tree Postorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        if not root: return []
        todo, nodes, last = [], [], None
        while root or todo:
            if root:
                todo.append(root)
                root = root.left
            else:
                node = todo[-1]
                if node.right and node.right != last:
                    root = node.right
                else:
                    nodes.append(node.val)
                    last = node
                    todo.pop()
        return nodes